src/Makefile.am

```diff
@@ -6,8 +6,112 @@
 
 DIST_SUBDIRS = secp256k1 univalue
 
+# This enables all warnings and treats them as errors. Individual warnings should be
+# 1. fixed,
+# 2. locally silenced with an explanatory comment, or
+# 3. globally silenced here with an explanatory comment.
+#
+# Justifications
+# • `-Wno-aix-compat`: we don’t want to maintain compatibility with IBM AIX compilers
+# • `-Wno-c++-compat`: this warns about an ABI layout incompatibility between C and C++,
+#    but we don’t need such compatibility except where we use `extern "C"`
+# • `-Wno-c++11-compat-pedantic`: we don’t want to maintain compatibility with older compilers
+# • `-Wno-c++14-compat-pedantic`: we don’t want to maintain compatibility with older compilers
+# • `-Wno-c++98-compat-pedantic`: we don’t want to maintain compatibility with older compilers
```

Comment:
`-Wc++98-compat-pedantic` and `-Wc++11-compat-pedantic` are composite options that include a grab-bag of things that are consistent with the "we don’t want to maintain compatibility with older compilers" rationale, and things that are not.

For example: [`-Wc++11-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-11-compat-pedantic) implies (among other things) [`-Wc++11-compat`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-11-compat). The latter category contains warnings about code that is likely to be incorrect in C++11 and later, even though the code might have been correct &mdash;or differently incorrect&mdash; in earlier versions. In most cases the code being warned about is invalid in C++11 and later. So as far as I can see, there's no advantage in disabling the `-Wc++11-compat` diagnostics because we "don’t want to maintain compatibility with older compilers"; quite the opposite. The same rationale applies to [`-Wc++98-compat`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-98-compat). (Not to be confused with [`-Wc++98-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-98-compat-pedantic)!)

[`-Wc++11-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wc-11-compat-pedantic) also implies [`-Wpre-c++20-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wpre-c-20-compat-pedantic) and [`-Wpre-c++2b-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wpre-c-2b-compat-pedantic), which we should not disable for a different reason: we are not using C++20 yet, and so we want to warn about code that looks like it might be using C++20 constructs that are not supported in C++17. (Just selecting `-std=c++17` doesn't necessarily do so, because the warning might be about code that was technically valid but will change behaviour.)

On the other hand, [`-Wpre-c++14-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wpre-c-14-compat-pedantic) and [`-Wpre-c++17-compat-pedantic`](https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html#wpre-c-17-compat-pedantic) contain diagnostics that we *don't* want (and so it is correct to use `-Wno-pre-c++14-compat-pedantic` and `-Wno-pre-c++17-compat-pedantic`, I think), because they will just stop us from using features of C++14 and C++17 (or earlier) that are perfectly harmless. These options don't have the same problem as `-Wc++11-compat-pedantic` or `-Wc++98-compat-pedantic`.

Frankly, the inconsistency in semantics of very similarly named options is a mess. The "pedantic" options [aren't documented for GCC](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html), so I don't know whether to blame GCC or clang devs for this, but whoever it was really dropped the ball.

Reply:
This is what I think we probably want:
```suggestion
# • `-Wno-c++98-compat`: we want to be able to use C++{11,14,17} constructs introduced after C++98
# • `-Wno-pre-c++14-compat-pedantic`: we want to be able to use C++14 constructs
# • `-Wno-pre-c++17-compat-pedantic`: we want to be able to use C++17 constructs
#
# Note that the naming of '-W*compat*' options is inconsistent; check
# <https://releases.llvm.org/15.0.0/tools/clang/docs/DiagnosticsReference.html> before
# changing this.
```

Reply:
In general, these aggregate flags are designed to be explicitly enabled, rather than disabled after `-Weverything`. So, yes, disabling them silences too many warnings[^1]. I think what we should do is re-enable the compatibility flags we want after disabling the ones we don’t. So here’s the pattern:

1. enable everything
2. disable `*-compat-pedantic` for C++ versions < current
3. enable `*-compat-pedantic` for C++ versions >= current

That (ignoring the flags not related to C++ versions) gets us to
```
-Weverything
-Wno-cxx98-compat-pedantic -Wno-c++11-compat-pedantic -Wno-c++14-compat-pedantic
-Wc++17-compat-pedantic -Wc++20-compat-pedantic
```

Personally, I like the consistency of that, rather than fiddling specific flags. We shouldn’t need `-Wc++17-compat-pedantic` at all (since those things should be errors in the version we’re on), but when we upgrade to C++20, we can switch `-Wc++17-compat-pedantic` to `-Wno` and should be good. And when Clang supports C++23, we add `-Wc++23-compat-pedantic` to the end of the list.

`-Wc++11-compat` _is_ an odd one, but I _think_ it’s ok. Its warnings are things that should be errors since we compile for C++17. E.g., “warning: ’_A_’ is a keyword in C++11” – we would never see that warning even if the flag is enabled (unless the keyword in question was removed in C++14/17), because we’d get an error complaining about _A_ being a keyword.

[^1]: It feels like there’s a variance issue here that is complicated by the order of flags on the command-line mattering. Without significant CLI order some other behavior might make more sense. E.g., enabling a superflag should enable all of its subflags (unless they’ve been explicitly disabled); but disabling a superflag should only disable subflags that aren’t enabled by some other superflag.

Reply:
I’ve made the change I suggested along with the move to configure.ac – but open to more discussion here.